### PR TITLE
SPE-4 Migrate & rework `dex.sandwiches`

### PR DESCRIPTION
**Pull request by hildobby**
# Thank you for contributing to Spellbook!
Please refer to the top of the `readme` in the root of Spellbook to learn how to contribute to Spellbook on DuneSQL.

### CODE REVIEW

**Comment by hildobby** [comment]
Huh "Table 'hive.dex.trades' does not exist"? What did I do wrong here @jeff-dude?

**Comment by Hosuke** [comment]
> Huh "Table 'hive.dex.trades' does not exist"? What did I do wrong here @jeff-dude?

This means `dex.trades` is not migrated yet(The spell does not have the `dunesql` tag).

We may use `dex.trades_migration_beta` instead for current testing.

**Comment by hildobby** [comment]
> trades_migration_beta

Alright thanks, I'll try out using `dex.trades_migration_beta` instead then

**Comment by jeff-dude** [comment]
> > trades_migration_beta
> 
> Alright thanks, I'll try out using `dex.trades_migration_beta` instead then

well, good news is `dex.trades` is now live on dunesql 🙌 

less good news, you'll have to change code back 🤷‍♂️

**Comment by hildobby** [comment]
> > > trades_migration_beta
> > 
> > 
> > Alright thanks, I'll try out using `dex.trades_migration_beta` instead then
> 
> well, good news is `dex.trades` is now live on dunesql 🙌
> 
> less good news, you'll have to change code back 🤷‍♂️

Damn that was quick wtf, I swear the migration started barely a week ago

**Comment by hildobby** [comment]
I'm really confused here, the schema it fails to find is the one it just created during tests? Do you know what's going on here @jeff-dude @0xRobin?

Screenshot 2023-09-22 at 11 22 28

File: macros/models/_sector/dex/dex_sandwiches.sql [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+, CASE WHEN s1.token_bought_address=s2.token_sold_address THEN 'token_sold' ELSE 'token_bought' END AS sandwiched_token
+, CASE WHEN s1.token_bought_address<s1.token_sold_address THEN 0 ELSE 1 END AS token_order
+FROM trades_with_index s1
+INNER JOIN trades_with_index s2 ON s1.block_number=s2.block_number
+    AND s1.project=s2.project
+    AND s1.version=s2.version
+    AND s1.tx_hash!=s2.tx_hash
+    AND s1.project_contract_address=s2.project_contract_address
+    --AND s2.token_sold_amount BETWEEN s1.token_bought_amount*0.9 AND s1.token_bought_amount*1.1
+INNER JOIN {{transactions}} tx1 ON  dt.blockchain='ethereum'
```
Comment (Hosuke) [inline macros/models/_sector/dex/dex_sandwiches.sql:35]: This line may be causing the problem.

Comment (hildobby) [reply macros/models/_sector/dex/dex_sandwiches.sql]: Ah good call, trying to remove it and seeing how it fares + some other small fixes

**Comment by couralex6** [comment]
@hildobby I was actually able to run this successfully locally. I'm suspecting the macros aren't being picked up by the CI for some reason. 

Let's merge this on monday when we have more people around in case we need to revert.

Thanks for migrating and refactoring this!

File: macros/models/_sector/dex/dex_sandwiches.sql [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    AND s1.version=s2.version
+    AND s1.tx_hash!=s2.tx_hash
+    AND s1.project_contract_address=s2.project_contract_address
+    AND s1.block_time >= date_trunc('day', now() - interval '3' day)
+    AND s2.block_time >= date_trunc('day', now() - interval '3' day)
+    {% if is_incremental() %}
+    AND s1.block_time >= date_trunc('day', now() - interval '7' day)
+    AND s2.block_time >= date_trunc('day', now() - interval '7' day)
+    {% endif %}
+INNER JOIN {{transactions}} tx1 ON dt.blockchain='ethereum'
```
Comment (Hosuke) [inline macros/models/_sector/dex/dex_sandwiches.sql:40]: By the way, I could not find the dt reference here.
Also do you think we may use the '{{blockchain}}' here?
```suggestion
INNER JOIN {{transactions}} tx1 ON dt.blockchain = '{{blockchain}}'
```

File: macros/models/_sector/dex/dex_sandwiches.sql [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+, tx1.index
+, s1.token_sold_amount_raw
+, s1.token_bought_amount_raw
+, s1.token_sold_amount
+, s1.token_bought_amount
+, CASE WHEN s1.tx_from=s2.tx_from THEN 'tx_from' ELSE 'taker' END AS commonality
+, CASE WHEN s1.token_bought_address=s2.token_sold_address THEN 'token_sold' ELSE 'token_bought' END AS sandwiched_token
+, CASE WHEN s1.token_bought_address<s1.token_sold_address THEN 0 ELSE 1 END AS token_order
+FROM trades_with_index s1
+INNER JOIN trades_with_index s2 ON s1.block_number=s2.block_number
```
Comment (0xRobin) [inline macros/models/_sector/dex/dex_sandwiches.sql:29]: Where does this raw `trades_with_index` come from?
I assume we'll want to use a chain specific dex_trades model that's passed in through the macro variable?

**Review by 0xRobin — CHANGES_REQUESTED** [review]

**Comment by 0xRobin** [comment]
> @hildobby I was actually able to run this successfully locally

curious what model you were able to run locally, as there seems to be still some issues in the macro that should make these models fail to run. :thinking:

**Comment by couralex6** [comment]
> > @hildobby I was actually able to run this successfully locally
> 
> @couralex6 curious what model you were able to run locally, as there seems to be still some issues in the macro that should make these models fail to run. 🤔

Oops you're right I was running on old version of the branch

**Comment by Hosuke** [comment]
Currently only sandwiches records on `ethereum` is available:
https://dune.com/queries/3056258

File: macros/models/_sector/dex/dex_sandwiches.sql [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+, s1.token_sold_amount_raw
+, s1.token_bought_amount_raw
+, s1.token_sold_amount
+, s1.token_bought_amount
+, s1.evt_index
+--, CASE WHEN s1.tx_from=s2.tx_from THEN 'tx_from' ELSE 'taker' END AS commonality
+--, CASE WHEN s1.token_bought_address=s2.token_sold_address THEN 'token_sold' ELSE 'token_bought' END AS sandwiched_token
+--, CASE WHEN s1.token_bought_address<s1.token_sold_address THEN 0 ELSE 1 END AS token_order
+FROM {{ ref('dex_trades') }} s1
+INNER JOIN {{ ref('dex_trades') }} s2 ON s1.blockchain='ethereum'
```
Comment (Hosuke) [inline macros/models/_sector/dex/dex_sandwiches.sql:30]: Maybe is due to this line?
```suggestion
INNER JOIN {{ ref('dex_trades') }} s2 ON s1.blockchain = '{{blockchain}}'
```

Comment (hildobby) [reply macros/models/_sector/dex/dex_sandwiches.sql]: Ah yeah ofc 😅 fixed it, thanks!

**Review by jeff-dude — APPROVED** [review]
beautiful PR here! love to see the use of macros becoming the norm, it makes everyone's lives easier.

thanks again for all the help in migration 🤝

**Pull request merged**